File: crates/agent-block/tests/fixtures/compile_loop_range_mock.lua
-- Fixture: compile_loop's read tools against a file that is too big to send.
--
-- Scenario (Anthropic mock, 3 iterations):
--   Iter 1: the mock calls fs_read on the whole file. It is over the size
--           threshold, so the answer is its length and a pointer to
--           read_file_range — no digest, no summarising sub-call.
--   Iter 2: the mock calls read_file_range(path, 10, 20) and gets those lines
--           verbatim and line-numbered, which is what fs_edit addresses by.
--   Iter 3: the mock calls fs_edit on the marker line (REPLACE_ME → DONE).
--           The verify then passes.
--
-- The runner runs after every one of those iterations, including the two that
-- only read: the verify is the loop's step and not the model's.
--
-- File structure (600 lines, ~15000 chars, well above the threshold):
--   Lines 1-9:   "-- pre-range line N"
--   Lines 10-20: "-- verbatim-line-NN" (unique prefix for the range assertion)
--   Lines 21-599: "-- padding line NNN"
--   Line 600:    "-- marker: REPLACE_ME"
--
-- The mock (spawn_range_mock) is launched by the Rust test fn and its base URL
-- is passed via ANTHROPIC_BASE_URL_TEST.
--
-- Prints READ_FILE_RANGE_VERBATIM_PASS on success.

local base_url = std.env.get("ANTHROPIC_BASE_URL_TEST")
assert(base_url, "ANTHROPIC_BASE_URL_TEST must be set")

local target_path = std.env.get("COMPILE_LOOP_RANGE_TARGET")
assert(target_path, "COMPILE_LOOP_RANGE_TARGET must be set")

-- Write the large target file (> 10000 chars to exceed the read threshold).
-- Lines 10-20 carry a distinctive prefix so the test can confirm verbatim content.
do
    local f = assert(io.open(target_path, "w"))
    for i = 1, 9 do
        f:write(string.format("-- pre-range line %d\n", i))
    end
    for i = 10, 20 do
        f:write(string.format("-- verbatim-line-%02d\n", i))
    end
    for i = 21, 599 do
        f:write(string.format("-- padding line %03d\n", i))
    end
    f:write("-- marker: REPLACE_ME\n")
    f:close()
end

local compile_loop = require("compile_loop")

-- mock_runner: passes once the marker line says DONE.
local runner_call_count = 0
local function mock_runner(paths)
    runner_call_count = runner_call_count + 1
    local path = type(paths) == "table" and paths[1] or paths
    local f = io.open(path, "r")
    if not f then
        return { ok = false, stderr = "cannot open " .. tostring(path), stdout = "", exit_code = 1 }
    end
    local content = f:read("*a") or ""
    f:close()
    if content:find("-- marker: DONE", 1, true) then
        return { ok = true, stdout = "DONE marker found", stderr = "", exit_code = 0 }
    end
    return { ok = false, stderr = "DONE marker not found in file", stdout = "", exit_code = 1 }
end

local td = compile_loop.make({
    runner = mock_runner,
    edit_mode = "diff",
    llm = {
        provider = "anthropic",
        base_url = base_url,
        api_key = "dummy",
        model = "mock-model",
    },
})

local result_json = td.handler({
    spec = "change the marker line from REPLACE_ME to DONE",
    target_files = { target_path },
})

local result = std.json.decode(result_json)
assert(result.ok, "compile_loop must converge using the range read, got: " .. tostring(result.summary or "?"))

-- The verify ran after every iteration, not only the one that edited.
assert(runner_call_count == 3, "the runner must run once per iteration, got " .. runner_call_count)

print("READ_FILE_RANGE_VERBATIM_PASS")

File: crates/agent-block/tests/common/compile_loop_range_mock.rs
//! In-process Anthropic mock for compile_loop's read tools against a file that
//! is too big to send.
#![allow(dead_code)]
//!
//! Three iterations, one model call each — the loop runs one beat per
//! iteration, so the turn is told apart by how many tool results the request
//! carries:
//!
//! - 0 results: `fs_read` on the whole file. It is over the size threshold, so
//!   the loop answers with the file's length and a pointer to
//!   `read_file_range`. (There is no digest and no summarising sub-call any
//!   more; this is what replaced them.)
//! - 1 result: `read_file_range(path, 10, 20)`, which must come back verbatim
//!   and line-numbered whatever the file's size.
//! - 2 results: `fs_edit` on the marker line, after which the verify passes.
//!
//! `MockState.tool_result_texts` records every tool result the model was shown,
//! in order, which is the only place a test can see what a tool handler
//! actually handed it.

use axum::{
    extract::State,
    http::{header, StatusCode},
    response::IntoResponse,
    routing::post,
    Router,
};
use serde_json::json;
use std::sync::{
    atomic::{AtomicUsize, Ordering},
    Arc, Mutex,
};

/// Shared state for the range mock server.
#[derive(Clone)]
pub struct MockState {
    /// Total HTTP requests received.
    pub call_count: Arc<AtomicUsize>,
    /// Every tool result text the mock has seen, in order of first appearance.
    pub tool_result_texts: Arc<Mutex<Vec<String>>>,
}

impl MockState {
    fn new() -> Self {
        Self {
            call_count: Arc::new(AtomicUsize::new(0)),
            tool_result_texts: Arc::new(Mutex::new(Vec::new())),
        }
    }

    /// Snapshot of the recorded tool result texts.
    pub fn tool_result_texts(&self) -> Vec<String> {
        self.tool_result_texts.lock().expect("lock").clone()
    }
}

/// The line the fixture puts the `REPLACE_ME` marker on.
const MARKER_LINE: u64 = 600;
const MARKER_EXPECT: &str = "-- marker: REPLACE_ME";
const MARKER_REPLACE: &str = "-- marker: DONE";

/// The `fs_edit` input that turns the marker line into `DONE`.
fn marker_edit_input(path: &str) -> serde_json::Value {
    json!({
        "path": path,
        "edits": [{
            "start_line": MARKER_LINE,
            "end_line": MARKER_LINE,
            "expect": MARKER_EXPECT,
            "replace": MARKER_REPLACE
        }]
    })
}

/// Tool result texts carried by an Anthropic request, in message order.
fn anthropic_tool_results(body: &serde_json::Value) -> Vec<String> {
    let mut out = Vec::new();
    let Some(messages) = body.get("messages").and_then(|m| m.as_array()) else {
        return out;
    };
    for msg in messages {
        if msg.get("role").and_then(|r| r.as_str()) != Some("user") {
            continue;
        }
        let Some(blocks) = msg.get("content").and_then(|c| c.as_array()) else {
            continue;
        };
        for b in blocks {
            if b.get("type").and_then(|t| t.as_str()) == Some("tool_result") {
                out.push(
                    b.get("content")
                        .and_then(|c| c.as_str())
                        .unwrap_or_default()
                        .to_string(),
                );
            }
        }
    }
    out
}

/// Record any tool result texts this request carries that we have not seen yet.
fn record_tool_results(state: &MockState, results: &[String]) {
    let mut seen = state.tool_result_texts.lock().expect("lock");
    for (i, text) in results.iter().enumerate() {
        if i >= seen.len() {
            seen.push(text.clone());
        }
    }
}

/// Extract the first absolute target file path from the request body.
///
/// The loop's opening message lists them:
/// ```text
/// Files:
///   /absolute/path/to/file
/// ```
fn extract_first_path(body: &serde_json::Value) -> Option<String> {
    let messages = body.get("messages").and_then(|m| m.as_array())?;
    for msg in messages {
        let content: String = {
            if let Some(s) = msg.get("content").and_then(|c| c.as_str()) {
                s.to_string()
            } else if let Some(blocks) = msg.get("content").and_then(|c| c.as_array()) {
                blocks
                    .iter()
                    .filter_map(|b| b.get("text").and_then(|t| t.as_str()))
                    .collect::<Vec<_>>()
                    .join("\n")
            } else {
                continue;
            }
        };
        let mut in_files = false;
        for line in content.lines() {
            if line.trim() == "Files:" {
                in_files = true;
                continue;
            }
            if in_files {
                let trimmed = line.trim();
                if trimmed.starts_with('/') {
                    return Some(trimmed.to_string());
                } else if !trimmed.is_empty() {
                    in_files = false;
                }
            }
        }
    }
    None
}

/// POST /v1/messages handler for the range test.
async fn anthropic_range_handler(
    State(state): State<MockState>,
    body: axum::body::Bytes,
) -> impl IntoResponse {
    let req_value = match serde_json::from_slice::<serde_json::Value>(&body) {
        Ok(v) => v,
        Err(e) => {
            eprintln!("[compile_loop_range_mock] bad request body: {e}");
            let err_body = json!({ "error": format!("bad request: {e}") }).to_string();
            return (
                StatusCode::BAD_REQUEST,
                [(header::CONTENT_TYPE, "application/json")],
                err_body,
            );
        }
    };

    let prev = state.call_count.fetch_add(1, Ordering::SeqCst);

    let results = anthropic_tool_results(&req_value);
    record_tool_results(&state, &results);
    let path = extract_first_path(&req_value).unwrap_or_else(|| "/unknown/path".to_string());

    let response_json = match results.len() {
        0 => json!({
            "id": format!("msg_range_read_{}", prev + 1),
            "type": "message",
            "role": "assistant",
            "content": [{
                "type": "tool_use",
                "id": "toolu_range_read_1",
                "name": "fs_read",
                "input": { "path": path }
            }],
            "model": "claude-haiku-mock",
            "stop_reason": "tool_use",
            "usage": { "input_tokens": 20, "output_tokens": 10 }
        }),
        1 => json!({
            "id": format!("msg_range_slice_{}", prev + 1),
            "type": "message",
            "role": "assistant",
            "content": [{
                "type": "tool_use",
                "id": "toolu_range_1",
                "name": "read_file_range",
                "input": { "path": path, "line_start": 10, "line_end": 20 }
            }],
            "model": "claude-haiku-mock",
            "stop_reason": "tool_use",
            "usage": { "input_tokens": 30, "output_tokens": 15 }
        }),
        2 => json!({
            "id": format!("msg_range_edit_{}", prev + 1),
            "type": "message",
            "role": "assistant",
            "content": [{
                "type": "tool_use",
                "id": "toolu_range_edit_1",
                "name": "fs_edit",
                "input": marker_edit_input(&path)
            }],
            "model": "claude-haiku-mock",
            "stop_reason": "tool_use",
            "usage": { "input_tokens": 40, "output_tokens": 20 }
        }),
        _ => json!({
            "id": format!("msg_range_done_{}", prev + 1),
            "type": "message",
            "role": "assistant",
            "content": [{ "type": "text", "text": "DONE" }],
            "model": "claude-haiku-mock",
            "stop_reason": "end_turn",
            "usage": { "input_tokens": 50, "output_tokens": 5 }
        }),
    };

    (
        StatusCode::OK,
        [(header::CONTENT_TYPE, "application/json")],
        response_json.to_string(),
    )
}

/// Spawn the mock on an ephemeral port.
///
/// # Returns
/// - `addr`: `SocketAddr`. Convert to a URL with `format!("http://{addr}")`.
/// - `state`: `call_count` should equal 3 after the subprocess, and
///   `tool_result_texts()` holds what each read handed the model.
///
/// # Panics
/// Panics only on OS-level port bind failure.
pub async fn spawn_range_mock() -> (std::net::SocketAddr, Arc<MockState>) {
    let state = Arc::new(MockState::new());

    let router = Router::new()
        .route("/v1/messages", post(anthropic_range_handler))
        .with_state((*state).clone());

    let listener = tokio::net::TcpListener::bind("127.0.0.1:0")
        .await
        .expect("bind ephemeral port for compile_loop range mock");
    let addr = listener.local_addr().expect("local_addr");

    tokio::spawn(async move {
        let _ = axum::serve(listener, router).await;
    });

    (addr, state)
}
